lib/DocumentRetriever.ts

```diff
@@ -179,7 +179,41 @@ DocumentRetriever.prototype.getRequest = async function(this: DocumentRetriever)
 		object.ExclusiveStartKey = Document.isDynamoObject(this.settings.startAt) ? this.settings.startAt : this.internalSettings.model.Document.objectToDynamo(this.settings.startAt);
 	}
 	if (this.settings.attributes) {
-		object.ProjectionExpression = this.settings.attributes.join(", ");
+		const expressionAttributeNames = {};
+		const projectionExpression = [];
+		// Check object ExpressionAttributeNames
+		if (object.ExpressionAttributeNames) {
+			let existingIndex = Object.keys(object.ExpressionAttributeNames).reduce((existing, item) => {
+				return Math.max(parseInt(item.replace("#a", "")), existing);
+			}, 0);
+			this.settings.attributes.forEach((element, index) => {
+				if (Object.keys(object.ExpressionAttributeNames).length) {
+					const item = Object.keys(object.ExpressionAttributeNames).find(key => object.ExpressionAttributeNames[key] === element);
+					if (item) {
+						projectionExpression.push(item);
+					} else {
+						const attrID = `#a${existingIndex + 1}`;
+						existingIndex++;
+						projectionExpression.push(attrID);
+						expressionAttributeNames[attrID] = element;
+					}
+				} else {
+					const attrID = `#a${index + 1}`;
+					projectionExpression.push(attrID);
+					expressionAttributeNames[attrID] = element;
+				}
+			});
+			object.ExpressionAttributeNames = { ...object.ExpressionAttributeNames, ...expressionAttributeNames };
+			object.ProjectionExpression = projectionExpression.sort().join(", ");
+		} else {
+			object.ExpressionAttributeNames = {};
+			this.settings.attributes.forEach((element, index) => {
+				const attrID = `#a${index}`;
+				projectionExpression.push(attrID);
+				object.ExpressionAttributeNames[attrID] = element;
+			});
+			object.ProjectionExpression = projectionExpression.sort().join(", ");
+		}
```

Comment:
I feel like a lot of this code overall can be reduced and made more streamlined. For example lines 195-198 are very similar to 201-203.

Also line 207 is the same as 215. So instead of putting those within the `if` & `else` blocks let's bring that down into the broader scope so we only have to run it once and it gets run regardless.

Let me know if you need further help with this. But I think there are a lot of ways to streamline this.